### ISSUE TITLE
Output layer, loss function and metrics

### ISSUE THREAD

**Issue by antonior92** [antonior92/physionet-12ecg-classification#17]
Hi everyone, 

With PR #15  I just included loading functions and the infrastructure for loading the new datasets, and also modified the last layer to deal with it. 

------

# The new metric

It is interesting to point out that the new metric (`challenge metric` in the plots bellow) completely change our goals. I quickly describe the new metric:
- Assume that you have `n_out` outputs, they compute `i = 1, ..., n_out` confusion matrices, one for each output. Lets call those `C_i`
- They comput an averaged confusion metric `C = (1 / n_out) * sum_i (C_i)` 
- They have a weights matrix `W` displayed bellow: 
![Screen Shot 2020-07-07 at 09 07 59](https://user-images.githubusercontent.com/16557411/86778983-6ba30e00-c031-11ea-920d-4c7b70bca5d5.png)
- They compute `Cw = sum(W * C)` where `*` is the element wise product
- They normalize `Cw` such that 1 correspond to getting all right diagnosis and 0 to guessing the normal diagnosis all the time. (Notice that is possible to get negative values)

---------
# Evaluating the new metric on the dataset/classes from the unofficial phase

So I started by evaluating the new metric in the dataset used in the unofficial phase dataset and considering only 8 classes. I compare four types of output layer:
-  Sigmoids in the output layer (all classes can occur at once);
- Softmax in the output layer (only one class can occur at once);
- A mix of sigmoid and softmax in the output layer encoding know mutually exclusive relations that I got from @tom1963br. I include two variation
   - The old implementation which used the softmax in the output layer but used threshold for its probabilities in the same way as we do for softmax;
   - A new and better implementation that choose only the highest probability in the case of softmax and threshold the softmax
In all cases, the output is corrected according to the occurrence  to account  for class imbalance

I display the result below, showing the challenge metric, together with f_beta, g_beta and validation loss:
![Screen Shot 2020-07-07 at 09 04 43](https://user-images.githubusercontent.com/16557411/86778603-f6373d80-c030-11ea-84ac-51a334ae31f6.png)

Notice that:
1. The validation loss increases after some point in training. Indicating a calibration problem. This problem does not reflect in the other metrics
2. The softmax+sigmoid (with the new implementation) yields the best results for `f_beta`, `g_beta` and the `challenge_metric`.
3. Just using a softmax yields better results regarding the challenge metric.
4. Just using the sigmoid yields the best results regarding `f_beta` and `g_beta`

----
# Evaluating the new metric on all datasets

Bellow, I compare the softmax and the sigmoid output layers, considering all datasets and all classes

### Softmax:
![Screen Shot 2020-07-07 at 09 05 06](https://user-images.githubusercontent.com/16557411/86778651-0222ff80-c031-11ea-8db7-b15992e15357.png)

------
### Sigmoid
![Screen Shot 2020-07-07 at 09 04 55](https://user-images.githubusercontent.com/16557411/86778615-fa635b00-c030-11ea-9e9e-82b74f77bd3a.png)

Notice:
1. The validation loss also increases, indicating a calibration problem. This problem does not reflect in the other metrics
2. Just using a softmax yields better results regarding the challenge metric.
3. Just using the sigmoid yields the best results regarding `f_beta` and `g_beta`
4. For sigmoid output, the `challenge metric` is negative. So, in some sense, it is worse than just predicting normal. 

----------

# Discussion

1. The new metric strongly penalizes the number of outputs `n_out`, while the old metric do not actually take that into account. So this is one of the reasons for the better results (regarding the `challenge_metric`) when using softmax. And the better `f_beta` and `g_beta` results are obtained when using sigmoid. 
2. `f_beta` and `g_beta` are largely invariant to the number of instance of each classe in the testset while the `challenge_metric` is now very much dependent of how many instances of each class are present in the test set. So an important question is: should we expect the hidden testset to be distributed as the datasets they provide us? In this case I think we should not correct for class imbalance during training anymore...
3. The experiment on the old dataset, show that encoding structure in the output layer works and allow us to get the best results in all metrics.

----
# Next steps

1. I believe we can explicitly penalize for `n_out` in the last layer by including some penalty on the L1 norm of the probability of the outputs (or something like that). This might allow us to use sigmoids with all classes with reasonably good results. I will see what I can do about that
2. As I said before, since the new metric depends strongly on the number of instance of each class during the evaluation, I believe we might need to remove any correction for the class imbalance. Since we might expect the same class imbalance during test. Might be a good idea to ask this on the challenge forum as well, and ask if the test set has the same class distribution as the training set.
3. Encoding structure in the last layer is important and we should try to encode mutually excussiveness and the classes hierarchy in the last layer.
